### ISSUE TITLE
Web [Charts]: Support label rotation and truncation

### ISSUE THREAD

**Issue by vovakulikov** [sourcegraph/sourcegraph#39032]
## Background

Currently, the bar chart (and all categorical-like charts) have a problem with label overlapping when dealing with many items (bars) datasets. See the image below.

<img width="394" alt="Screenshot 2022-07-19 at 15 24 18" src="https://user-images.githubusercontent.com/18492575/179752725-7f16444d-d023-44aa-8a6e-74b0a6083689.png">

_Code Insights compute live preview bar chart_

**We should** 
- Render only the appropriate number of legend items based on their length and available space
- Rotate labels to increase square for label rendering
- Truncate label text if they are too long to fit in available label space

**Comment by sourcegraph-bot-2** [comment]
Heads up @joelkw @felixfbecker @vovakulikov @unclejustin - the "team/code-insights" label was applied to this issue.

**Comment by vovakulikov** [comment]
@felixfbecker I'm not sure that even truncate labels will help here without rotation logic, to be honest. In compute we usually have a lot of groups/bars, so we always have to deal with a large number of labels, which means most of the labels will be truncated to just 1-2 symbols, which isn't so helpful. 

Currently, users can still see groups' names through the tooltip; considering that this is the first version of compute-powered insight, I think this should be fine for now. But I agree that this looks awful, and this issue should be prioritised if we want to keep going in different chart types directions or support the second versions of compute-powered insight. cc @Joelkw 

Also, for the record, there wasn't a quick fix for it. At least I didn't find it to include this in the 3.42.

**Comment by felixfbecker** [comment]
@vovakulikov pulling this out from Slack:

> Worked today on chart labels problem, And it feels like there is no a good way to achieve a reliable and consistent label layout across all major browsers (this time Safari turned out has a few problems with g element) in all corner cases. I had a few thoughts on this but probably I will capture them in the charts doc and just fix cases that we're going to use for the bar chart only.

How do other charting libraries (vsix, recharts, highcharts, ...) do this? This feels like it should be a solved problem where we shouldn't have to reinvent the wheel and implement a solution we came up with all by ourselves for the whole matrix of browsers.

Could you share your thoughts/the charts doc you're referring to?

**Comment by vovakulikov** [comment]
> This feels like it should be a solved problem where we shouldn't have to reinvent the wheel and implement a solution we came up with all by ourselves for the whole matrix of browsers.

I was surprised but this problem isn't a wheel at all. Most chart libraries just ignore this problem or provide non optimal solutions without label layout changes. Like visx for example simple [says](https://github.com/airbnb/visx/issues/1493#issuecomment-1119854986) that consumers should know about possible width of ticks and adjust margin manually. They have text helpers though but this doesn't solve the problem in general. Especially for us where we will deal with big labels (I suspect this is true based on my expires with data that we usually have from compute aggregations). The only one library that I've seen that does tick layout rotation based on tick width is https://d3fc.io but it doesn't have a good React support so we can't just use it and also I don't want to bring another tool that solves only one specific problem.

I think I found a solution that covers most of our cases with layouts and possible long labels. But it requires a significant change in how we're building charts at the moment from components API perspective. This is why I mentioned a chart doc, in my head this doc is supposed to be a main entry point if you want to build a chart in Sourcegraph where we will describe all tech/UI/a11y parts of data visualisation components that we have. 

We had to do it from the very beginning but we didn't have time for it, we had just a few components and we didn't have enough chart consumers. It seems that we have enough counts in all three directions now to pay more attentions to the documentation of chart UI.
